### PR TITLE
Feat: Transformative flavour

### LINKED ISSUE #8: Context flavour does not respect adjusted return types of `ctx.reply`
# Feature request

If I adjust the return type of `ctx.reply` using a different plugin, such as the hydrate plugin, I would like the type of `ctx.replyWithEmoji` to adjust accordingly. This makes it necessary to use a transformative context flavour, rather than an additive one.

### PR DESCRIPTION
**Pull request by andrrms**
Changes made, needs review.
If merged, plugin version will go to 2.0.0, as it also changes the name of the flavoring interface.
Fix #8

### CODE REVIEW

File: src/mod.ts [hunk 1]
```diff
@@ -2,7 +2,8 @@ import { Context, Message, Middleware } from "./deps.deno.ts";
 import getEmoji from "./emoji.ts";
 import type { EmojiName } from "./emoji.ts";
 
-export interface EmojiFlavor {
+export type EmojiFlavour<C extends Context> = Flavour & C;
```
Comment (KnorpelSenf) [inline src/mod.ts]: This is BrE, please use AmE (`Flavor` rather than `Flavour`)

**Review by KnorpelSenf — CHANGES_REQUESTED** [review]

**Comment by KnorpelSenf** [comment]
This does not fix #8. This PR would need to extract the return type of `ctx.reply` and use it in `ctx.replyWithEmoji`.

**Comment by andrrms** [comment]
Major version change will not be required anymore.

**Comment by KnorpelSenf** [comment]
Note that you're admin in the repo, there's no need to open pull requests from the fork. You can just open them from other branches in the same repo :)

**Review by KnorpelSenf — COMMENTED** [review]
This still does not solve the issue.

The aim is to make the following code compile:

```ts
import { Bot } from "https://deno.land/x/grammy/mod.ts";
import {
    hydrate,
    HydrateFlavor,
} from "https://deno.land/x/grammy_hydrate/mod.ts";

type MyContext = EmojiFlavor<HydrateFlavor<Context>>;

const bot = new Bot<MyContext>("");

bot.use(hydrate(), emojiParser(), async (ctx) => {
    const first = await ctx.reply("Ehh");
    await first.delete();
    const second = await ctx.replyWithEmoji`Hi! ${"robot"}`;
    await second.delete();
});
```
However, I cannot get it to work myself. It seems like `ReturnType` somehow swallows the changes that the inner transformative context flavour is performing to `ctx.reply`. Hence, the emoji flavour does not pick this up so it cannot provide the same type signature as `ctx.reply` has.

I do not see how this can be fixed. Feel free to investigate. Perhaps the hydrate plugin needs to be adjusted, too, who knows.

**Comment by KnightNiwrem** [comment]
> This still does not solve the issue.
> 
> The aim is to make the following code compile:
> 
> ```ts
> import { Bot } from "https://deno.land/x/grammy/mod.ts";
> import {
>     hydrate,
>     HydrateFlavor,
> } from "https://deno.land/x/grammy_hydrate/mod.ts";
> 
> type MyContext = EmojiFlavor<HydrateFlavor<Context>>;
> 
> const bot = new Bot<MyContext>("");
> 
> bot.use(hydrate(), emojiParser(), async (ctx) => {
>     const first = await ctx.reply("Ehh");
>     await first.delete();
>     const second = await ctx.replyWithEmoji`Hi! ${"robot"}`;
>     await second.delete();
> });
> ```
> 
> However, I cannot get it to work myself. It seems like `ReturnType` somehow swallows the changes that the inner transformative context flavour is performing to `ctx.reply`. Hence, the emoji flavour does not pick this up so it cannot provide the same type signature as `ctx.reply` has.
> 
> I do not see how this can be fixed. Feel free to investigate. Perhaps the hydrate plugin needs to be adjusted, too, who knows.

It doesn't appear to be _just_ `ReturnType`, but rather a behavior of Typescript, itself. It seems like function type intersections are essentially treated as overloads, and Typescript will simply pick one of them for the return value type (without much consistency).

[Typescript Playground](https://www.typescriptlang.org/play?#code/JYWwDg9gTgLgBAbzgIQjANHGBPMBTOAYQgDsY8APeAXzgDMoIQ4ByAcygEMQRsWBuAFCCc+OAFk4AXiKlyVANoAiKHjAAbbAHVgMABYBBAK4ATYBCUBdOADJZZSjGWqN2K7fvynKtZp36DElBOGHMSKyFRAgAlaThovBgjKBIAFVw8AB5xAD4hQRM8AGN1TlU4ItIAZ3gYAC4JIUqSGrgoOJgACiUlAEohIA)

**Comment by andrrms** [comment]
> Note that you're admin in the repo, there's no need to open pull requests from the fork. You can just open them from other branches in the same repo :)

I do it on impulse, I'm still learning to use git 😅

**Comment by KnorpelSenf** [comment]
> It seems like function type intersections are essentially treated as overloads

This is correct, it makes a lot of sense that this is the root cause of the issue.

I'm still unsure how to proceed from here.

**Comment by KnightNiwrem** [comment]
@andrrms @KnorpelSenf Please take a look at https://github.com/grammyjs/hydrate/pull/1/files if it may be a potential workaround.

**Comment by KnorpelSenf** [comment]
@andrrms can you please check out https://github.com/grammyjs/hydrate/pull/1 and confirm that everything works for you?

**Pull request closed without merging**